### ISSUE TITLE
Mirror Pod cannot be updated / upgraded

### ISSUE THREAD

**Issue by dchen1107** [GoogleCloudPlatform/kubernetes#6580]
I upgraded my cluster to use the latest etcd image gcr.io/google_containers/etcd:2.0.9 by using push.sh:
```
root@kubernetes-master:/var/log# docker ps
CONTAINER ID        IMAGE                                                COMMAND                CREATED             STATUS              PORTS               NAMES
2752cfc55b8c        gcr.io/google_containers/etcd:2.0.9                  "/usr/local/bin/etcd   22 seconds ago      Up 21 seconds                           k8s_etcd-container.1e717455_etcd-server-kubernetes-master_default_5f35d49b23c06deb8fddf656fa1003b6_3dde77d4            
```

The cluster successfully upgraded, but I found error message in kubelet.log: 
```
E0407 23:32:03.532257   18134 kubelet.go:1335] Failed creating a mirror pod "etcd-server-kubernetes-master_default": &errors.StatusError{ErrStatus:api.Status{TypeMeta:api.TypeMeta{Kind:"", APIVersion:""}, ListMeta:api.ListMeta{SelfLink:"", ResourceVersion:""}, Status:"Failure", Message:"pods \"etcd-server-kubernetes-master\" already exists", Reason:"AlreadyExists", Details:(*api.StatusDetails)(0xc208474c00), Code:409}}
```

Because the mirrored Pod cannot updated, kubectl get pods shows wrong information:

```
etcd-server-kubernetes-master                                         etcd-container          gcr.io/google_containers/etcd:2.0.8                  kubernetes-master.c.golden-system-455.internal/                       <none>                                                              Running             23 hours
```

Please note that above output still indicates it is 2.0.8, not 2.0.9. 

cc/ @yujuhong

**Comment by yujuhong** [comment]
@dchen1107, do we have any restrictions on static pod updates? Should we allow arbitrary fields to be updated (kubelet doesn't validate updates of static pods, but apiserver does).

**Comment by dchen1107** [comment]
How complicated to add validate updates to kubelet? The current problem is that the mirror ones doesn't reflect the real state of those static pods. For example, we use whatever our deployment tool to upgrade our master daemons running in a pod, such as more newly release version, or resource spec, the mirror ones should reflect such changes.

**Comment by yujuhong** [comment]
There is a simple way to *reflect* the static pod changes: delete the mirror pod and re-create it.

Instead, if we send an update request to the apiserver, the apiserver may reject the update if it fails the validation. My question is that do we expect static pods to be updated arbitrarily?

**Comment by dchen1107** [comment]
No, I think we should follow the same update rules for static pods as regular ones.

**Comment by dchen1107** [comment]
I filed #6594 for more design discussion on static pods versioning. For v1.0, let's simply delete/re-creation the mirror ones when static pods are updated.

**Comment by yujuhong** [comment]
There are pros and cons for each approach (recreate #6607 vs update #6607).

I uploaded PR #6604 originally to send an actual update request, and surface errors through events if apiserver rejects the update. However, after uploading it I found this may be confusing sometimes because kubelet's inability to tell between pod delete/recreate vs pod updates.

For example, if user *removes* a manifest file, and later decide to add the file back with a completely different spec. This would be considered *valid* by the apiserver because the mirror pod would effectively be deleted and then re-created. However, if the user does this operation in a small window of time , kubelet may think this is an update and refuse it. This would get worse if we add in that kubelet may restart, etc.

Also, mirror pod exists to reflect the true status of the static pod. If kubelet already accepts the change, there is no reason not to reflect this change via the mirror pod.

In short, recreation seems like a better idea to me (PR #6607).

**Comment by bgrant0607** [comment]
Recreation seems fine.

**Comment by vmarmol** [comment]
@yujuhong this is fixed as of #6607 no?

**Comment by yujuhong** [comment]
Yes, closing.
